File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionSubstring.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;

public class TransformerFunctionSubstring extends TransformerFunction {
    public TransformerFunctionSubstring() {
        super(FunctionDescription.of(
            Map.of(
            "begin", ArgumentType.of(ArgType.Integer).position(0).defaultInteger(0),
            "end", ArgumentType.of(ArgType.Integer).position(1).defaultIsNull(true)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var str = context.getString(null);
        if (str == null) {
            return null;
        }
        var length = str.length();
        var beginIndex = context.getInteger("begin");
        if (beginIndex == null) {
            return str;
        }
        if (beginIndex < 0) {
            beginIndex = Math.max(0, length + beginIndex);
        }
        var endValue = context.getInteger("end");
        if (endValue == null) {
            return str.substring(beginIndex);
        }
        var endIndex = Math.min(endValue, length);
        if (endIndex < 0) {
            endIndex = Math.max(0, length + endIndex);
        }
        return str.substring(beginIndex, endIndex);
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionSubstring.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import JsonElementStreamer from "../JsonElementStreamer";

class TransformerFunctionSubstring extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        begin: { type: ArgType.Integer, position: 0, defaultInteger: 0 },
        end: { type: ArgType.Integer, position: 1, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const str = await context.getString(null);
    if (str == null) {
      return null;
    }
    let beginIndex = await context.getInteger("begin");
    if (beginIndex == null) {
      return str;
    }

    const endIndex = await context.getInteger("end");
    return str.slice(beginIndex, endIndex ?? undefined);
  }
}

export default TransformerFunctionSubstring;
